langchain/src/tools/outlook/authFlowREST.ts

```diff
@@ -0,0 +1,205 @@
+import * as http from "http";
+import * as url from "url";
+import { AuthFlowBase } from "./authFlowBase.js";
+import { getEnvironmentVariable } from "../../util/env.js";
+
+interface AccessTokenResponse {
+  access_token: string;
+  refresh_token: string;
+}
+
+// this class uses your clientId, clientSecret and redirectUri
+// to get access token and refresh token, if you already have them,
+// you can use AuthFlowToken or AuthFlowRefresh instead
+export class AuthFlowREST extends AuthFlowBase {
+  private clientSecret: string;
+
+  private redirectUri: string;
+
+  private port: number;
+
+  private pathname: string;
+
+  private refreshToken = "";
+
+  constructor({
+    clientId,
+    clientSecret,
+    redirectUri,
+  }: { clientId?: string; clientSecret?: string; redirectUri?: string } = {}) {
+    let id = clientId;
+    let secret = clientSecret;
+    let uri = redirectUri;
+    if (!id || !secret || !uri) {
+      id = getEnvironmentVariable("OUTLOOK_CLIENT_ID");
+      secret = getEnvironmentVariable("OUTLOOK_CLIENT_SECRET");
+      uri = getEnvironmentVariable("OUTLOOK_REDIRECT_URI");
+    }
+    if (!id || !secret || !uri) {
+      throw new Error("Missing clientId, clientSecret or redirectUri.");
+    }
+    super(id);
+    this.clientSecret = secret;
+    this.redirectUri = uri;
+    const parsedUrl = new URL(this.redirectUri);
+    this.port = parsedUrl.port ? parseInt(parsedUrl.port, 10) : 3000;
+    this.pathname = parsedUrl.pathname || "";
+  }
+
+  // Function to open or print the login url for user to login
+  private openAuthUrl(): string {
+    const loginEndpoint =
+      "https://login.microsoftonline.com/common/oauth2/v2.0/authorize";
+    const { clientId } = this; // client ID regestered in Azure
+    const response_type = "code";
+    const response_mode = "query";
+    const redirectUri = encodeURIComponent(this.redirectUri); // redirect URI regestered in Azure
+    const scope = encodeURIComponent(
+      "openid offline_access https://graph.microsoft.com/.default"
+    );
+    const state = "12345";
```

Comment:
Double checking this is the correct value?

Reply:
Just found that this is optional so I will remove this